src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.Analysis.Tree.cs

```diff
@@ -0,0 +1,505 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+using System.Linq;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    internal partial class LambdaRewriter
+    {
+        internal sealed partial class Analysis : BoundTreeWalkerWithStackGuardWithoutRecursionOnTheLeftOfBinaryOperator
+        {
+            /// <summary>
+            /// This is the core node for a Scope tree, which stores all semantically meaningful
+            /// information about declared variables, closures, and environments in each scope.
+            /// It can be thought of as the essence of the bound tree -- stripping away many of
+            /// the unnecessary details stored in the bound tree and just leaving the pieces that
+            /// are important for closure conversion. The root scope is the method scope for the
+            /// method being analyzed and has a null <see cref="Parent" />.
+            /// </summary>
+            [DebuggerDisplay("{ToString(), nq}")]
+            private sealed class Scope
+            {
+                public Scope Parent { get; }
+
+                public ArrayBuilder<Scope> NestedScopes { get; } = ArrayBuilder<Scope>.GetInstance();
+
+                /// <summary>
+                /// A list of all closures (all lambdas and local functions) declared in this scope.
+                /// </summary>
+                public ArrayBuilder<Closure> Closures { get; } = ArrayBuilder<Closure>.GetInstance();
+
+                /// <summary>
+                /// A list of all locals or parameters that were declared in this scope and captured
+                /// in nested scopes. "Declared" refers to the start of the variable lifetime (which,
```

Comment:
Captured in nested scopes or also this one? Assuming it means this scope and nested scopes.